lib/active_shipping/carriers/usps.rb

```diff
@@ -336,6 +408,171 @@ def canned_address_verification_works?
       xml.at('/CarrierPickupAvailabilityResponse/City').try(:text) == 'SAN FRANCISCO' && xml.at('/CarrierPickupAvailabilityResponse/Address2').try(:text) == '18 FAIR AVE'
     end
 
+    def build_shipment_request(action, origin, destination, package, options = {})
+      if us_address?(destination)
+        build_us_shipment_request(action, origin, destination, package, options)
+      else
+        build_world_shipment_request(action, origin, destination, package, options)
+      end
+    end
+
+    def build_us_shipment_request(action, origin, destination, package, options = {})
+      xml_builder = Nokogiri::XML::Builder.new do |xml|
+        xml.send(XML_ROOTS[action].to_sym, 'USERID' => @options[:login]) do
+          service = SHIPMENT_SERVICES[options[:service]]
+
+          unless service
+            raise ArgumentError, "Service is not provided or not supported."
+          end
+
+          build_us_location_node(xml, 'From', origin)
+          build_us_location_node(xml, 'To', destination)
+
+          xml.WeightInOunces("%0.1f" % [package.ounces, 1].max)
+          xml.ServiceType(service)
+
+          xml.ImageType(LABEL_TYPE)
+
+          size_code = USPS.size_code_for(package)
+          container = CONTAINERS[package.options[:container]]
+          container ||= (package.cylinder? ? 'NONRECTANGULAR' : 'RECTANGULAR') if size_code == 'LARGE'
+          xml.Container(container)
+
+          xml.Size(size_code)
+        end
+      end
+
+      save_request(xml_builder.to_xml)
+    end
+
+    def build_us_location_node(xml, prefix, address)
+      array = [
+        ['Name', address.name],
+        ['Firm', address.company],
+        ['Address1', address.address1],
+        ['Address2', address.address2 || address.address1],
```

Comment:
I don't follow. Looks like `ActiveShipping::Location#address1` is exactly what we need here for `Address1`, since `ActiveShipping::Location#address2` contains additional information about that address in cases you're referring to.